src/darwin.rs

```diff
@@ -148,6 +162,117 @@ pub fn get_cpu_time(val: &rusage) -> f64 {
+
+    // io_name_t is defined as an array of 128 chars.
+    // https://developer.apple.com/documentation/kernel/io_name_t
+    let mut name = [0u8; 128];
+
+    let entry = 'entry: {
+        loop {
+            // SAFETY: Even if iter is invalid, this is safe to call, and will return 0.
+            let entry = unsafe { IOIteratorNext(iter) };
+            if entry == 0 {
+                break;
+            }
+            // SAFETY: name is io_name_t, which has size of 128 chars. It remains alive until
+            // the end of this function.
+            if map_mach_resp(unsafe { IORegistryEntryGetName(entry, name.as_mut_ptr().cast()) }).is_err() {
+                // SAFETY: An error occurred so we must release the current entry.
+                unsafe {
+                    IOObjectRelease(entry);
+                    continue;
+                }
+            }
+
+            // IORegistryEntryName always returns a C-string name assigned to a registry entry,
+            // if there is no error.
+            // https://developer.apple.com/documentation/iokit/1514323-ioregistryentrygetname
+            let name = CStr::from_bytes_until_nul(&name).expect("kernel iterator should always return null-terminator");
+            if name.to_bytes() == b"pmgr" {
+                break 'entry entry;
+            } else {
+                // SAFETY: We are finished with the entry and should release it.
+                unsafe {
+                    IOObjectRelease(entry);
+                }
+            }
+        }
+
+        return Err(SporkError::unimplemented());
+    };
+
+    let p_core_ref = unsafe {
+        // SAFETY: CFStringCreateWithBytesNoCopy does not accept null-terminators,
+        // and does not need to be deallocated since we're using the static string
+        // as the backing memory.
+        let prop_name = CFStringCreateWithBytesNoCopy(
+            kCFAllocatorDefault,
+            VOLTAGE_STATE5_SRAM.as_ptr(),
+            VOLTAGE_STATE5_SRAM.len() as CFIndex,
+            CFStringBuiltInEncodings_kCFStringEncodingUTF8,
+            0,
+            kCFAllocatorNull,
+        );
+
+        // SAFETY: CFStringCreateWithBytesNoCopy is infallible, so prop_name is always valid.
+        IORegistryEntryCreateCFProperty(entry, prop_name, kCFAllocatorDefault, 0)
+    };
+
+    if p_core_ref.is_null() {
+        return Err(SporkError::unimplemented());
+    }
+
+    /// SAFETY: p_core_ref has been checked for null.
+    let p_core_len = unsafe { CFDataGetLength(p_core_ref.cast()) };
+    if p_core_len < 8 {
+        return Err(SporkError::unimplemented());
+    }
+
+    let range = CFRange {
+        location: p_core_len - 8,
+        length: 4,
+    };
+
+    let mut cpu_freq = [0u8; 4];
+    // SAFETY: p_core_len must be at least 8 bytes. Therefore, the
+    // 4 byte range requested, from 8 bytes before the end of the buffer,
+    // always falls within the buffer of the data p_core_ref points to.
```

Comment:
Unfortunately Apple doesn't give official documentation for this, the thread I linked was basically people reverse engineering the output of the data from the IO registry. The reason why the length of the data must be at least 8 bytes is because that is the minimum for one "power state" definition which is 2 4-byte integers, first of which corresponds to a frequency, and the second *probably* corresponding with the power state level.

To summarize, we can look at the raw data with `ioreg`

```
ioreg -l | rg voltage-states5-sram
    | |   | |   "voltage-states5-sram" = <0046c3230003000000475a31000300000048f13e1903000000643f4d350300000065d65a4503000000666d6858030000006704766703000000689b838d0300000084e991c50300000085809f01040000008617ad2c04000000aff5b42c0400000087aeba2c04000000d8d3bc2c040000005f67c02c040000>
```

Splitting that raw data up into 4 byte integers we get (little endian)

```
0046c323 00030000
00475a31 00030000
0048f13e 19030000 
00643f4d 35030000 
0065d65a 45030000 
00666d68 58030000 
00670476 67030000 
00689b83 8d030000 
0084e991 c5030000 
0085809f 01040000 
008617ad 2c040000
00aff5b4 2c040000 
0087aeba 2c040000 
00d8d3bc 2c040000 
005f67c0 2c040000
```

The first number in this array (0x23c34600) is 600000000 = 600MHz. The second number (0x00000300) is less than 1 MHz so it can't be a frequency. Hence it seems to be sorted in order of increasing frequency. The second to last number (8 byte offset from the end) is 0xc0675f00 = 3228000000 = 3.228 GHz.

Some more discussion on this [here](https://www.github.com/giampaolo/psutil/pull/2222#discussion_r1161142859) but basically it also matches up with what the commenter finds on their M2 laptop.